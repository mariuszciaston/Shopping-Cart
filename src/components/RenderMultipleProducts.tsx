import { ProductsContextType } from "@/types";
import { Link } from "react-router-dom";

const RenderMultipleProducts: React.FC<ProductsContextType> = ({
  productsData,
  error,
  loading,
}) => {
  return (
    <>
      {loading && <p>Loading...</p>}
      {error && <p>A network error was encountered</p>}

      <div className="grid grid-cols-[repeat(auto-fill,minmax(240px,1fr))] content-start gap-4 sm:gap-8">
        {productsData &&
          productsData.map((product) => (
            <Link to={`/products/${product.id}`} key={product.id}>
              <div className="flex h-full flex-col gap-4 rounded-md bg-white p-4 shadow-md">
                <img
                  src={product.image}
                  alt={product.title}
                  className="aspect-square bg-white object-contain"
                />
                <h2 className="font-bold">{product.title}</h2>
                <p>${product.price}</p>
              </div>
            </Link>
          ))}
      </div>
    </>
  );
};

export default RenderMultipleProducts;
