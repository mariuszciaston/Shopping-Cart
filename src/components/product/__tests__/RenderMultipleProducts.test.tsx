import { render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import RenderMultipleProducts from "../RenderMultipleProducts";
import { describe, it, expect } from "vitest";

describe("RenderMultipleProducts", () => {
  const products = [
    {
      id: 1,
      title: "Product 1",
      price: 100,
      image: "image1.jpg",
      description: "Description for Product 1",
    },
    {
      id: 2,
      title: "Product 2",
      price: 200,
      image: "image2.jpg",
      description: "Description for Product 2",
    },
  ];

  it("should display loading message when loading is true", () => {
    render(
      <RenderMultipleProducts productsData={[]} loading={true} error={null} />,
    );
    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  it("should display error message when error is true", () => {
    render(
      <RenderMultipleProducts
        productsData={[]}
        loading={false}
        error={"error message"}
      />,
    );
    expect(
      screen.getByText("A network error was encountered"),
    ).toBeInTheDocument();
  });

  it("should display products when data is available", () => {
    render(
      <BrowserRouter>
        <RenderMultipleProducts
          productsData={products}
          loading={false}
          error={null}
        />
      </BrowserRouter>,
    );

    products.forEach((product) => {
      const allProductLinks = screen.getAllByRole("link");
      const productLink = allProductLinks.find(
        (link) => link.getAttribute("href") === `/products/${product.id}`,
      );
      expect(productLink).not.toBeNull();
      expect(productLink).toHaveAttribute("href", `/products/${product.id}`);

      const productImg = screen.getByAltText(product.title);
      expect(productImg).toHaveAttribute("src", product.image);

      expect(screen.getByText(product.title)).toBeInTheDocument();
      expect(screen.getByText(`$${product.price}`)).toBeInTheDocument();
    });
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <BrowserRouter>
        <RenderMultipleProducts
          productsData={products}
          loading={false}
          error={null}
        />
      </BrowserRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
