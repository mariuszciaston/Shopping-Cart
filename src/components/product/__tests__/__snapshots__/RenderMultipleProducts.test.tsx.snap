// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RenderMultipleProducts > matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="grid grid-cols-[repeat(auto-fill,minmax(240px,1fr))] content-start gap-4 sm:gap-8"
  >
    <a
      data-discover="true"
      href="/products/1"
    >
      <div
        class="flex h-full flex-col gap-4 rounded-md bg-white p-4 shadow-md"
      >
        <img
          alt="Product 1"
          class="aspect-square bg-white object-contain"
          src="image1.jpg"
        />
        <h2
          class="font-bold"
        >
          Product 1
        </h2>
        <p>
          $100
        </p>
      </div>
    </a>
    <a
      data-discover="true"
      href="/products/2"
    >
      <div
        class="flex h-full flex-col gap-4 rounded-md bg-white p-4 shadow-md"
      >
        <img
          alt="Product 2"
          class="aspect-square bg-white object-contain"
          src="image2.jpg"
        />
        <h2
          class="font-bold"
        >
          Product 2
        </h2>
        <p>
          $200
        </p>
      </div>
    </a>
  </section>
</DocumentFragment>
`;
